### PR TITLE
KAFKA-17542: Use actions/labeler for automatic PR labeling - part 2

### PR DESCRIPTION
**Pull request by TaiJuWu**
Jira: https://issues.apache.org/jira/browse/KAFKA-17542
test on: https://github.com/TaiJuWu/kafka/pull/11

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: .github/workflows/labeler.yml [hunk 1]
```diff
@@ -29,3 +28,39 @@ jobs:
     - uses: actions/labeler@v5
       with:
         configuration-path: .github/configs/labeler.yml
+    - uses: actions/checkout@v4
+      with:
+        ref: ${{ github.event.pull_request.head.sha }}
+    - name: Compare trunk and PR
```
Comment (chia7712) [inline .github/workflows/labeler.yml]: How about renaming it to "check MINOR label"?

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for this patch. what about the plan of "tests" label?

File: .github/workflows/labeler.yml [hunk 2]
```diff
@@ -29,3 +28,39 @@ jobs:
...
+      env:
+        GH_TOKEN: ${{ github.token }}
+      run: |
+        BASE_COMMIT="${{ github.event.pull_request.base.sha }}"
+        HEAD_COMMIT="${{ github.event.pull_request.head.sha }}"
+
+        summary=$(git diff --stat $BASE_COMMIT $HEAD_COMMIT | tail -n 1)
```
Comment (mumrah) [inline .github/workflows/labeler.yml]: Not sure if it helps, but we can use the `gh` CLI here. 

https://cli.github.com/manual/gh_pr_diff

**Review by mumrah — COMMENTED** [review]
@TaiJuWu, thanks for following up with this patch!

Can we create this as a standalone bash script that lives in .github/scripts? I don't like having too much code directly in the workflow definition. This also makes it easier to test locally.

The script inputs (the SHAs to compare) can be env variables. If we need outputs, we can use the stdout "FOO=BAR" notation (see https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-output-parameter). 

A few other considerations:

1) Should we have other size labels? E.g., "tee-shirt" sizes: small, medium, large, xlarge
2) If a PR's size changes after an update, we may need to remove an existing label
3) Do we need to exclude any files from this diff sizing? I can't think of any...

**Comment by TaiJuWu** [comment]
> @TaiJuWu thanks for this patch. what about the plan of "tests" label?

Hi @chia7712 , the test label are supported in last PR [here](https://github.com/TaiJuWu/kafka/blob/e39f883572faaaf6cdcb76d0fa1d00ebb3db1795/.github/configs/labeler.yml#L29)
I think it meets your requirement.

**Comment by TaiJuWu** [comment]
> @TaiJuWu, thanks for following up with this patch!
> 
> Can we create this as a standalone bash script that lives in .github/scripts? I don't like having too much code directly in the workflow definition. This also makes it easier to test locally.

Sure.
> 
> The script inputs (the SHAs to compare) can be env variables. If we need outputs, we can use the stdout "FOO=BAR" notation (see https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-output-parameter).
> 

Thanks for information!
> A few other considerations:
> 
> 1. Should we have other size labels? E.g., "tee-shirt" sizes: small, medium, large, xlarge
> 2. If a PR's size changes after an update, we may need to remove an existing label
> 3. Do we need to exclude any files from this diff sizing? I can't think of any...

I completely agree. To support multiple size labels, we can use [PR-size-labeler](https://github.com/CodelyTV/pr-size-labeler?tab=readme-ov-file#%EF%B8%8F-arguments). However, a limitation is that it requires us to set exactly five labels.

**Comment by mumrah** [comment]
> To support multiple size labels, we can use [PR-size-labeler](https://github.com/CodelyTV/pr-size-labeler?tab=readme-ov-file#%EF%B8%8F-arguments)

Please note that while we can use marketplace actions, we must have them approved by ASF Infra team unless it's from GitHub or a verified developer. I think this is a simple enough action that we can roll our own, plus it gives us more flexibility for the labels (e.g., we can start with just "small")

**Comment by apoorvmittal10** [comment]
> test on: [TaiJuWu#10](https://github.com/TaiJuWu/kafka/pull/10)

I don't find any changes as per this PR on mentioned test PR, am I missing something?

**Comment by TaiJuWu** [comment]
> > test on: [TaiJuWu#10](https://github.com/TaiJuWu/kafka/pull/10)
> 
> I don't find any changes as per this PR on mentioned test PR, am I missing something?

Hi @apoorvmittal10 , thanks for review.
This tes PR only shows the action can label `mirror` automatically as below.
image

By the way, we may use another tool help us to finish this target, you can refere [here](https://github.com/apache/kafka/pull/17260#issuecomment-2372267502)

**Comment by TaiJuWu** [comment]
> > To support multiple size labels, we can use [PR-size-labeler](https://github.com/CodelyTV/pr-size-labeler?tab=readme-ov-file#%EF%B8%8F-arguments)
> 
> Please note that while we can use marketplace actions, we must have them approved by ASF Infra team unless it's from GitHub or a verified developer. I think this is a simple enough action that we can roll our own, plus it gives us more flexibility for the labels (e.g., we can start with just "small")

I file a jira for this https://issues.apache.org/jira/browse/INFRA-26158

**Comment by mumrah** [comment]
@TaiJuWu I thought we were going to stick with your script for this? The PR-size-labeler is going to add a size label to every PR, which I'm not sure we want. I like the approach you started in this PR for just labeling "minor" PRs

**Comment by TaiJuWu** [comment]
> @TaiJuWu I thought we were going to stick with your script for this? The PR-size-labeler is going to add a size label to every PR, which I'm not sure we want. I like the approach you started in this PR for just labeling "minor" PRs

Oh, sorry for the misunderstanding. 
If we only need a mirror label, we will have to write our own script, as the `PR-size-labeler` cannot be used in this case. 
I'm also not entirely sure which solution is the best for us. We can start with just the `mirror` and consider using the `PR-size-labeler` in the future if necessary.

**Comment by mumrah** [comment]
> Oh, sorry for the misunderstanding.

No worries 😄 . I think we can start with your script and update it as needed. 

So, to finish this up, we need to

1) Move the bash into a script
2) Change the label name to `minor` (not `mirror`)
3) Address @chia7712's question about the "test" label. If this is too complex we can create a new JIRA for it and work on it later

File: .github/workflows/labeler.yml (hunk 2 above)
Comment (TaiJuWu) [reply .github/workflows/labeler.yml]: Thanks. It is very useful.

**Comment by TaiJuWu** [comment]
> > Oh, sorry for the misunderstanding.
> 
> No worries 😄 . I think we can start with your script and update it as needed.
> 
> So, to finish this up, we need to
> 
> 1. Move the bash into a script
> 2. Change the label name to `minor` (not `mirror`)
> 3. Address @chia7712's question about the "test" label. If this is too complex we can create a new JIRA for it and work on it later

Hi @mumrah ,
[1] and [2] are finished. I will check [3] with @chia7712 to ensure that the current `test label` is functioning properly.

**Review by mumrah — COMMENTED** [review]
Thanks for the updates, the script looks good.

I think we should considered a different label than `minor`. After all, we may have very serious PRs which only have a few changes :). "Minor" can also convey importance or significance. I suggest we go with "small" which is more accurate.

**Comment by TaiJuWu** [comment]
> Thanks for the updates, the script looks good.
>
>I think we should considered a different label than minor. After all, we may have very serious PRs which >only have a few changes :). "Minor" can also convey importance or significance. I suggest we go with >"small" which is more accurate.

Thanks for review and suggestion, updated!

**Review by mumrah — APPROVED** [review]
LGTM, thanks @TaiJuWu!

**Comment by chia7712** [comment]
@TaiJuWu thanks for this useful feature!

File: .github/scripts/label_small.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+LABEL_NAME=small
+MAX_SIZE=100
+
+pr_diff=$(gh pr diff $PR_NUM -R $GITHUB_REPOSITORY)
+
+insertions=$(printf "$pr_diff" | grep '^+' | wc -l)
```
Comment (chia7712) [inline .github/scripts/label_small.sh:22]: It seems that printf tries to parse the content of pr_diff, which may lead to invalid formats (see #17338).

Perhaps we should avoid manually parsing the content and use gh pr view instead. For example:

```
pr_diff=$(gh pr view $PR_NUM -R $GITHUB_REPOSITORY --json additions,deletions)

insertions=$(echo "$pr_diff" | jq -r '.insertions')
deletions=$(echo "$pr_diff" | jq -r '.deletions')

total_changes=$((insertions + deletions))
```

@TaiJuWu WDYT? could you open a MINOR to fix it?

Comment (TaiJuWu) [reply .github/scripts/label_small.sh:22]: Thanks for your remind, will update.

**Pull request merged**